File: stonex_ops/tools/probe_connections.py
"""probe_connections: run connection probes for a tenant or all tenants.

The stonx CLI owns the probe catalog. stonex-ops intentionally passes through
all returned connection IDs, including provider, freshness, report, AI, and
mail readiness checks.

This is the ONLY tool that triggers `stonx ctl probe`, which may write
`ops.connection_probe_state` through stonx.

Read-only — does not send notifications.
"""

from __future__ import annotations

import json
import sys
from typing import Any, Optional

from stonex_ops.executor import execute
from stonex_ops.whitelist import Probe, ProbeAll

_EMPTY_RESULT: dict[str, Any] = {
    "scope": None,
    "scopes": [],
    "result": "failed",
    "check_count": 0,
    "problem_count": 0,
    "connections": [],
}


async def run(
    stonx_bin: str,
    env: str,
    path: str,
    tenant: Optional[str] = None,
    shop_id: Optional[str] = None,
) -> dict[str, Any]:
    if tenant:
        cmd = Probe(tenant=tenant, shop_id=shop_id)
    else:
        cmd = ProbeAll()

    output = await execute(stonx_bin, env, path, cmd)

    try:
        parsed = json.loads(output.stdout)
    except json.JSONDecodeError:
        print(
            f"stonex-ops: probe returned invalid JSON "
            f"(length={len(output.stdout)}): {output.stdout[:300]!r}",
            file=sys.stderr,
        )
        error = output.stderr.strip() or "stonx probe produced no JSON output"
        return {**_EMPTY_RESULT, "error": error}

    # Collect connections from all scopes, injecting tenant/shop_id context.
    all_connections: list[dict[str, Any]] = []
    for scope in parsed.get("scopes", []):
        for conn in scope.get("connections", []):
            entry = dict(conn)
            if "tenant" in scope:
                entry["tenant"] = scope["tenant"]
            if "shop_id" in scope:
                entry["shop_id"] = scope["shop_id"]
                entry["shop_label"] = scope.get("shop_label", scope["shop_id"])
            all_connections.append(entry)

    scope_info = None
    if tenant:
        scope_info = {"tenant": tenant, "shop": shop_id}

    return {
        "scope": scope_info,
        "scopes": parsed.get("scopes", []),
        "result": parsed.get("result"),
        "check_count": parsed.get("check_count"),
        "problem_count": parsed.get("problem_count"),
        "connections": all_connections,
    }

File: src/tools/probe_connections.rs
//! probe_connections:探测一个租户(或所有租户)的数据连接状态。
//!
//! 支持的 probe 类型:
//! - all:所有连接 (xhs-api, jst/wangdian, runtime, freshness)
//! - xhs-api:小红书 API 连接
//! - jst:JST ERP 连接
//! - wangdian:旺店通 ERP 连接
//! - runtime:运行时健康
//! - freshness:数据新鲜度

use serde::Serialize;
use std::path::PathBuf;

use crate::executor;
use crate::whitelist::AllowedCommand;

#[derive(Debug, Serialize)]
pub(crate) struct ProbeConnectionsResult {
    pub scope: Option<ProbeScope>,
    pub result: Option<String>,
    pub checks: Option<usize>,
    pub problems: Option<usize>,
    pub connections: Vec<serde_json::Value>,
}

#[derive(Debug, Serialize)]
pub(crate) struct ProbeScope {
    pub tenant: Option<String>,
    pub shop: Option<String>,
}

pub(crate) async fn run(
    stonx_bin: PathBuf,
    env: &str,
    path: &str,
    tenant: Option<&str>,
    shop_id: Option<&str>,
) -> Result<ProbeConnectionsResult, String> {
    let cmd = match tenant {
        Some(tenant) => AllowedCommand::Probe {
            tenant: Some(tenant.to_string()),
            shop_id: shop_id.map(|s| s.to_string()),
        },
        None => AllowedCommand::ProbeAll,
    };

    let output = executor::execute(stonx_bin, env, path, &cmd)
        .await
        .map_err(|e| format!("probe failed: {e}"))?;

    let parsed: serde_json::Value =
        serde_json::from_str(&output.stdout).map_err(|e| format!("parse probe json: {e}"))?;

    // 收集所有 scope 的 connections
    let mut all_connections = Vec::new();
    if let Some(scopes) = parsed.get("scopes").and_then(|v| v.as_array()) {
        for scope in scopes {
            if let Some(conns) = scope.get("connections").and_then(|v| v.as_array()) {
                for conn in conns {
                    let mut entry = conn.clone();
                    // 注入 scope 信息
                    if let Some(obj) = entry.as_object_mut() {
                        if let (Some(tenant), Some(shop)) = (
                            scope.get("tenant").and_then(|v| v.as_str()),
                            scope.get("shop").and_then(|v| v.as_str()),
                        ) {
                            obj.insert("tenant".to_string(), serde_json::json!(tenant));
                            obj.insert("shop".to_string(), serde_json::json!(shop));
                        }
                    }
                    all_connections.push(entry);
                }
            }
        }
    }

    Ok(ProbeConnectionsResult {
        scope: tenant.map(|t| ProbeScope {
            tenant: Some(t.to_string()),
            shop: shop_id.map(|s| s.to_string()),
        }),
        result: parsed.get("result").and_then(|v| v.as_str()).map(|s| s.to_string()),
        checks: parsed.get("checks").and_then(|v| v.as_u64()).map(|v| v as usize),
        problems: parsed.get("problems").and_then(|v| v.as_u64()).map(|v| v as usize),
        connections: all_connections,
    })
}
